src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,33 +645,31 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
-            ODataPathSegment lastSegment = segments.Last();
+            int lastIndex = segments.Count - 1;
+            ODataPathSegment lastSegment = segments[lastIndex];
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                lastSegment = segments[lastIndex--];

```

Comment:
>lastIndex-- [](start = 39, length = 11)

Shouldn't this be --lastIndex?  I.e., don't we need to decrement *before* using the value?

Reply:
I went ahead and made this change, as well as the similar change when looping through TypeSegments, and removed the redundant lastSegment=segments[lastIndex].

---
In reply to: [480401681](https://github.com/OData/odata.net/pull/1854#discussion_r480401681) [](ancestors = 480401681)